File: 2962-count-subarrays-where-max-element-appears-at-least-k-times/2962-count-subarrays-where-max-element-appears-at-least-k-times.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        mv = max(nums)
        res = st = mv_win = 0

        for end in nums:
            if end == mv:
                mv_win += 1
            while mv_win==k:
                if nums[st] == mv:
                    mv_win -= 1
                st += 1
            res += st
        return res

File: 2962-count-subarrays-where-max-element-appears-at-least-k-times/2962-count-subarrays-where-max-element-appears-at-least-k-times.rs
impl Solution {
    pub fn count_subarrays(nums: Vec<i32>, k: i32) -> i64 {
        let n = nums.len();
        let mv = *nums.iter().max().unwrap();
        let mut res = 0;
        let mut cnt = 0;
        let mut idx = Vec::new();
        let k = k as usize;

        for i in 0..n {
            if nums[i] == mv {
                idx.push(i as i64);
                cnt += 1;
            }
            if cnt >= k {
                res += idx[idx.len() - k] + 1;
            }
        }

        res
    }
}
